README.md
# ceInstall

Installation and environment settings of the following libraries:

- banks
- ccdb
- clhep
- evio
- geant4
- geant4Data
- gemc
- jana
- mlibrary
- qt
- root
- scons
- xercesc


To release a JLAB_VERSION: 

1. change release date in ce/jlab.csh script
2. change default JLAB_VERSION in jlab.csh and jlab.sh (usually from devel to tag).  Change (production) on softenv.*
3. change default JLAB_VERSION in go_reload
4. make sure versions.env is updated
5. if necessary add the QT_VERSION tag in softenv.csh and jlab.sh for JLAB.
6. make sure the geant4 data versions are correct, in geant4.env AND in go_geant4
7. add the tag in sbsinstall/template alternative installation lines. Notice you need to add that line in devel_install as that one is not overwritten.
8. Use go_tar ceInstall tag local. 
9. test installation, copy log in sbsInstall/template_install.html
10. test installation on docker
11. after test add github tag. Notice: deleting a github release is not enough: need to delete the tag also.
12. tag the libraries below. To tag, use v1.4 and 1.4 as release name. Add notes.

Remember that go_tar supports the "local" third option to create a "debug" tar file instead of using the github tag feature.

- banks
- gemc 
- mlibrary
- scons_bm


Location to copy the tar.gz files:

/u/group/12gev_phys/www/packages/sources

NOTES:

- go_tar ceInstall will generate the _install.html based on the release number
- for JLAB interactive farm. Usually it cannot access the outside web, but setting these bypass this:

a. setenv http_proxy http://jprox.jlab.org:8082
b. setenv https_proxy http://jprox.jlab.org:8082

For ROOT in particular one can also download these "src" and put them in the "dst".

a. src1='http://lcgpackages.web.cern.ch/lcgpackages/tarFiles/sources/lz4-v1.7.5.tar.gz' to: dst1='/group/halla/analysis/rootbuild/ifarm1402/root-6.12.06/build/LZ4-prefix/src/lz4-v1.7.5.tar.gz'
b. src2='http://lcgpackages.web.cern.ch/lcgpackages/tarFiles/sources/tbb2017_U5.tar.gz' to: dst2='/group/halla/analysis/rootbuild/ifarm1402/root-6.12.06/build/TBB-prefix/src/tbb2017_U5.tar.gz'



TODO:

- add CERN LIBS if user so requests


changelog
----------------
 List of changes
 ---------------

7/12/19

- copying mysql files instead of making links:
  The files are small and this solve the problem of a central installation and user mounting that installation from remote computer cause the link would point to their local /usr/lib


6/11/19
- added support for newest clang 

4/22/19
- fixed go_root disabling packages
- changed compilation log screen size
- fixed devel cases for gemc and mlibrary
- devel has devel version for banks, gemc, mlibrary and ceinstall
- found optimal (hopefully) configuration for root

4/19/19
- added ccdb notes for ssl
- devel version with new geant4, clhep
- fixed go_reload with correct argument
- added xterm command with compilation log
- fixed documentation
- changed released date
- move ccdb ssl note to go_ccdb
- mlibrary log fix
- reformatting geant4 data and adding 2 missing tables 

.. Note::
	10/02/18 ceInstall tag 2.3

|

10/02/18
- fixed get_coptions logFile and logFileS if needed else ifs
- xercesc version increased to 3.2.2. Splitting env lib/lib64 check for Darwin/Linux
- go_field changed to reflect latest magnetic field maps. 

9/28/18
- fixed go_qt for Linux CentOS flavor
- fixed go_evio log
- fixed go_reload for 2.3
- added  -Wno-dev to suppress cmake warnings
- fixed softenv.sh typo

9/27/18
- gemc moved to github release mechanism
- at JLAB QT_VERSION is 5.10.1 not the system
- added geant4 data default env and utils to print out settings in printVersion.csh
- updated jana package name
- 2.3 release finalized
- better log of softenv

9/25/18
- removed 32 bit support
- preparing Announcement for 2.3
- setting wget command is different on Linux or Darwin. Moving -qc and show-progress options to a shell variable mwget
- xercesc: in linux, linking lib64 to lib. Why oh why they keep doing this.
- fixes to go_qt script offline installer, and go_geant4 repo name
- added systematic way of getting geant4 data
- fixed data retrievel for geant4
- added log of ROOT installation path

9/24/18
- formatting go_tar exclude from tar
- better README and walkthrough explanation
- better step-by-step instruction formatting
- added qc option to wget: quiet and do not retrieve if file exists already (avoids writing duplicates)
- added doc for default system installation of qt
- using $getCoptions throughout the installation scripts
- added wget --show-progress option
- added support for custom list of software library: only the packages in .jlab_software will be considered in the environment
- getting ready to release 2.3

8/28/18
- qt system is now the default mode. DEFAULT_QT_VERSION is set to system, and go_qt checks QT_SYSTEM
- clhep env sets the path
- jana env sets the path and plugin path

4/11/18
- compilation procedure standardized for package gemc
- compilation procedure standardized for package root
- root version for devel
- compilation procedure standardized for package banks
- compilation procedure standardized for package jana

4/10/18
- added utility functions to get build log (and its shortcut name)
- added utility function to log the start and ending of the compilation
- quiet option to src wget
- removed 32 bit options checks
- compilation procedure standardized for package clhep
- compilation procedure standardized for package xercesc
- added geant4 copy script
- removed clang_option
- compilation procedure standardized for package geant4
- added local scons_bm to packaging
- compilation procedure standardized for package evio
- compilation procedure standardized for package ccdb
- compilation procedure standardized for package mlibrary


3/29/18
- added note on instructions
- modified go_mlibrary to include LIBRARY="shared" option, in case users want to build gemc as a library (adds -fPic)

.. Note::
	3/21/18 ceInstall tag 2.2

|

3/21/18
- fixed cmake version check script

3/12/18
- adding prototype for system qt in Linux distributions.
- reversing order of cadmesh / scons so include files can be found
- adding local version of mlibrary to release
- updated instructions
- a bit of colors in the instructions
- fixed go_qt endif missing
- added timing for various packages
- limiting number of parallel compilation in mlibrary
- fixed go_banks untarring

3/11/18
- silencing untar with options instead of redirection
- using git release mechanism for scons_bm
- can use system qt if necessary
- fixed ncpu command
- updated announcement, untar command, instructions for devel and 2.2
- added templated installation instructions so we do not need to keep different versions of all
- fix go_tar / go_reload with proper paths, adding local option to go_tar
- mlibrary going to git releases for devel as well
- mlibrary release 1.3 instead of 1.2 as git did not include "include" in 1.2

3/9/18
 - portable command to get number of cpus
 - cadmesh release made part of mlibrary (but it will still git-get the submodules assimp and tetgen)
 - using git release mechanism for mlibrary and banks

3/7/18
 - starting release of 2.2

11/17/17
 - added ROOT includes

11/7/17
 - added gmedia in xcode headers

11/2/17
 - added gdata in xcode headers

10/31/17
 - added ghit in xcode headers

10/5/17
 - fixed uncommented comment
 - added CLANG cmake options for Darwin for geant4 and cmake

3/24/17
- added mt option for installation
- added clhep and geant4 in xcode settings

1/26/17
- introducing two python functions to streamline and standardize the versioning and user customization options

11/9/16
- limiting max parallel compilation to 24
  just in case there are local limits (for example, some JLab nodes)

7/22/16
- removed obsolete go_scrat
- finalized go_mlibrary
- added mlibrary in go_tar

5/12/16
- fixed BANKS envs

4/8/16
 - go_geant4 fixes
 - JLAB_ROOT not necessary when using keepmine

4/6/15
 - fixed deleting directories in go_geant4

4/5/16
 - added warnings for xercesc and qt

4/1/16
 - fixed overwriting of the env

3/31/16

ce/osrelease.pl
#!/usr/bin/env perl

use warnings;
use strict;

# Generate a string specific to the current posix platform for
# purposes of separating and distributing binaries.
#
# The output will have the form:
#
#  PosixType_OSVersion##-architecture-compilerType##
#
# Some examples:
#
# Linux_CentOS6.5-x86_64-gcc4.4.7
# Darwin__macosx10.11-x86_64-clang7.0.2
# Linux_Fedora9-i686-gcc4.3.0
#
# Errors are printed to stderr and some attempt is made to return
# a reasonable string for the system such that the script
# will never fail.
#

my $uname = `uname`;
chomp $uname;

my $compilerVersion = 'unknown';
my $release = 'unknown';


### Linux ########################
# Assuming gcc as default compiler
# --------------------------------
if ($uname eq 'Linux')
{
	### Fedora
	# time to make these general
	if (-e '/etc/fedora-release')
	{
		my     $release_string = `cat /etc/fedora-release`;
		if    ($release_string =~ /^Fedora release 28.*/)     { $release = '_Fedora28'; }
		elsif ($release_string =~ /^Fedora release 27.*/)     { $release = '_Fedora27'; }
		elsif ($release_string =~ /^Fedora release 26.*/)     { $release = '_Fedora26'; }
		elsif ($release_string =~ /^Fedora release 25.*/)     { $release = '_Fedora25'; }
		elsif ($release_string =~ /^Fedora release 24.*/)     { $release = '_Fedora24'; }
		elsif ($release_string =~ /^Fedora release 23.*/)     { $release = '_Fedora23'; }
		elsif ($release_string =~ /^Fedora release 22.*/)     { $release = '_Fedora22'; }
		elsif ($release_string =~ /^Fedora release 21.*/)     { $release = '_Fedora21'; }
		else
		{
			print STDERR "unrecognized Fedora release\n";
			$release = '_Fedora';
		}
	}
	### Red-Hat, SL, CentOS
	elsif (-e '/etc/redhat-release')
	{
		my $releaseString = `cat /etc/redhat-release`;
		if    ($releaseString =~ /^CentOS.*/)     { $release = 'CentOS'; }
		elsif ($releaseString =~ /^Red.*/)        { $release = 'RedHat'; }
		elsif ($releaseString =~ /^Scientific.*/) { $release = 'SL'; }
		
		# get everything after release
		my @releaseN = split /release/, $releaseString;
		# there may be a version in parenthesis
		my @releaseNs = split /\(/, $releaseN[1];
		# getting just the number
		my $relVersion = $releaseNs[0];
		# remove spaces at the beginning and end
		$relVersion =~ s/^\s+|\s+$//g;
		$release = $release.$relVersion;
	}
	### Ubuntu
	elsif (-e '/etc/lsb-release')
	{
		my $distrib_id = `cat /etc/lsb-release | grep DISTRIB_ID`;
		$distrib_id =~ s/DISTRIB_ID=//;
		my $distrib_release = `cat /etc/lsb-release | grep DISTRIB_RELEASE`;
		$distrib_release =~ s/DISTRIB_RELEASE=//;
		chomp $distrib_id;
		chomp $distrib_release;
		$release = "_${distrib_id}${distrib_release}";
	}
	###
	else
	{
		print STDERR "unrecognized Linux release\n";
		$release = 'linuxUnknown';
	}
	my $compV = qx(gcc -dumpversion);
	# chomp it to remove carriage return
	chomp $compV;
	$compilerVersion = "gcc".$compV;
}

### Mac ###################
# Clang is default compiler
# -------------------------
elsif ($uname eq 'Darwin')
{
	my $release_string = `uname -r`;
	if    ($release_string =~ /^13.*/) { $release = 'macosx10.9'; }
	elsif ($release_string =~ /^14.*/) { $release = 'macosx10.10'; }
	elsif ($release_string =~ /^15.*/) { $release = 'macosx10.11'; }
	elsif ($release_string =~ /^16.*/) { $release = 'macosx10.12'; }
	elsif ($release_string =~ /^17.*/) { $release = 'macosx10.13'; }
	elsif ($release_string =~ /^18.*/) { $release = 'macosx10.14'; }
	else
	{
		print STDERR "unrecognized Mac OS release: $release_string\n";
		$release = 'macosxUnknown';
	}
	
	
	# the output of clang -v is to stderr so redirecting output to stdout
	my $clangV = qx( clang -v 2>&1 | head -1);
	# removing first words from outout
	$clangV =~ s/Apple LLVM version//;
	$clangV =~ s/Apple clang version//;
	# getting just the version number
	my @clangN = split " ", $clangV;
	$compilerVersion = "clang".$clangN[0];
}

# This part sets the processor type and compiler version number
my $processor = qx(uname -m);
# chomp it to remove carriage return
chomp $processor;


print "${uname}_${release}-${processor}-${compilerVersion}\n";
exit;












